### ISSUE TITLE
clippy::explicit_counter_loop false positive when loop counter is not usize

### ISSUE THREAD

**Issue by lopopolo** [rust-lang/rust-clippy#4139]
```console
$ cargo clippy -V
clippy 0.0.212 (fc96aa03 2019-05-04)
```

I am iterating over a `Vec` and calling into a C binding that takes an `i64` index so I am using a loop counter to avoid a fallible conversion from `usize` instead of using `Vec::len`.

```rust
impl FromMrb<Vec<Value>> for Value {
    type From = Rust;
    type To = Ruby;

    fn from_mrb(interp: &Mrb, value: Vec<Self>) -> Self {
        let array = unsafe { sys::mrb_ary_new(interp.borrow().mrb) };
        let mut idx = 0;
        for item in value.into_iter() {
            // sys::mrb_ary_set takes an i64 index
            unsafe { sys::mrb_ary_set(interp.borrow().mrb, array, idx, item.inner()) };
            idx += 1;
        }
        Self::new(interp, array)
    }
}

```

```
error: the variable `idx` is used as a loop counter.
  --> mruby/src/convert/array.rs:49:21
   |
49 |         for item in value.into_iter() {
   |                     ^^^^^^^^^^^^^^^^^ help: consider using: `for (idx, item) in value.into_iter().enumerate()`
   |
note: lint level defined here
  --> mruby/src/lib.rs:1:9
   |
1  | #![deny(clippy::all, clippy::pedantic)]
   |         ^^^^^^^^^^^
   = note: #[deny(clippy::explicit_counter_loop)] implied by #[deny(clippy::all)]
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#explicit_counter_loop
```

**Comment by Manishearth** [comment]
I don't consider this a false positive, you can and perhaps should be casting here.

That you don't want to is a whole other thing: if you don't want to you should turn the lint off for this code.

**Comment by lopopolo** [comment]
Thanks @Manishearth. I have no trouble disabling the lint, but found it unexpected which is why I filed the ticket. Thanks for taking a look!
